### PR TITLE
gh-124552 : Improve the accuracy of possible breakpoint check in bdb

### PR DESCRIPTION
**Pull request by gaogaotiantian**
Now we check whether the line number is in the code object with `co_lines()` so we can have an accurate answer.

A bug needs to be fixed for generators. This is not caused by the new feature, but exposed. I put those together because the changes are very small. I can separate those if that's desired.

* Issue: gh-124552

### CODE REVIEW

**Comment by iritkatriel** [comment]
This needs a test.

File: Lib/bdb.py [hunk 1]
```diff
@@ -275,7 +280,25 @@ def do_clear(self, arg):
...
+    def lineno_in_frame(self, lineno, frame):
+        """Return True if the line number is in the frame's code object.
+        """
+        code = frame.f_code
+        if lineno < code.co_firstlineno:
+            return False
+        if code not in self.code_lineno:
+            self.code_lineno[code] = set()
+            for _, _, lineno in code.co_lines():
+                self.code_lineno[code].add(lineno)
```
Comment (iritkatriel) [inline Lib/bdb.py]:
```suggestion
            self.code_lineno[code] = set([lno for _, _, lno in code.co_lines()])
```

**Comment by gaogaotiantian** [comment]
Do you think a test that directly tests `break_anywhere` in `test_bdb` would work? It's a public documented API. Checking whether the dispatch function stopped would be a bit more complicated.

**Comment by gaogaotiantian** [comment]
I added a test for `break_anywhere`. Also I used the generator for `set()` instead of a list.

**Comment by terryjreedy** [comment]
I am testing this with IDLE's bdb-based debugger now.

**Comment by terryjreedy** [comment]
No (new) problems found.  Breakpoints at top level and in def still work.

**Comment by gaogaotiantian** [comment]
Thanks @terryjreedy for confirming this. #124533 might be another interesting one to test.

File: Lib/bdb.py [hunk 2]
```diff
@@ -275,7 +280,23 @@ def do_clear(self, arg):
...
+
+    def lineno_in_frame(self, lineno, frame):
+        """Return True if the line number is in the frame's code object.
+        """
+        code = frame.f_code
+        if lineno < code.co_firstlineno:
+            return False
+        if code not in self.code_lineno:
+            self.code_lineno[code] = set(lineno for _, _, lineno in code.co_lines())
+        return lineno in self.code_lineno[frame.f_code]
```
Comment (iritkatriel) [inline Lib/bdb.py]:
```suggestion
        return lineno in self.code_lineno[code]
```

File: Lib/bdb.py [hunk 3]
```diff
@@ -36,6 +37,7 @@ def __init__(self, skip=None):
         self.frame_returning = None
         self.trace_opcodes = False
         self.enterframe = None
+        self.code_lineno = weakref.WeakKeyDictionary()
```
Comment (iritkatriel) [inline Lib/bdb.py]:
```suggestion
        self.code_linenos = weakref.WeakKeyDictionary()
```

File: Misc/NEWS.d/next/Library/2024-09-25-22-06-52.gh-issue-124552.1nQKNM.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Improve the accuracy of possible breakpoint check in bdb so we can disable unnecessary events in functions.
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Library/2024-09-25-22-06-52.gh-issue-124552.1nQKNM.rst]:
```suggestion
Improve the accuracy of :mod:`bdb`'s check for the possibility of breakpoint in a frame. This makes it possible to disable unnecessary events in functions.
```

File: Lib/bdb.py [hunk 5]
```diff
@@ -275,7 +280,23 @@ def do_clear(self, arg):
     def break_anywhere(self, frame):
         """Return True if there is any breakpoint for frame's filename.
```
Comment (iritkatriel) [inline Lib/bdb.py]: this comment needs updating (it's not just filename now).

File: Lib/bdb.py [hunk 6]
```diff
@@ -155,6 +157,9 @@ def dispatch_return(self, frame, arg):
         if self.stop_here(frame) or frame == self.returnframe:
             # Ignore return events in generator except when stepping.
             if self.stopframe and frame.f_code.co_flags & GENERATOR_AND_COROUTINE_FLAGS:
+                # It's possible to trigger a StopIteration exception in
+                # the caller so we must set the trace function in the caller
+                self._set_caller_tracefunc(frame)
```
Comment (iritkatriel) [inline Lib/bdb.py:162]: how is this related to this PR?

Comment (gaogaotiantian) [reply Lib/bdb.py:162]: There is a hidden bug which was exposed by the change of `break_anywhere`. There's a test case of `bdb` testing raising `StopIteration` in a generator. It passed because `break_anywhere` always return `True` when the function was defined in the same file, which always sets the trace function on the caller because it is in the same file. That coincidence hide the bug where returning from a generator should stop in the caller. There's a similar call in line 177 - that's for the normal return case from https://github.com/python/cpython/pull/118979 - it's also my fix to a similar case.

**Review by iritkatriel — APPROVED** [review]
